src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java

```diff
@@ -250,18 +250,34 @@ public Flavor getFlavor() {
     }
 
     /**
-     * @return true IFF the receiver is a reference projection of a primitive class type and false
-     * for primitives or plain references
+     * @return true IFF the receiver is a reference projection type of a *value favoring* primitive class
+     * and false otherwise. Or in other words, is this a class type non-redundantly notated with .ref ??
      */
     public boolean isReferenceProjection() {
         return false;
     }
 
     /**
-     * @return the value projection type IFF the receiver is a reference projection of a primitive class type
-     * and null otherwise
+     * @return true IFF the receiver is a primitive reference type and false otherwise.
+     */
+    public boolean isPrimitiveReferenceType() {
+        return false;
+    }
+
+    /**
+     * @return true IFF the receiver is a value projection of a *reference favoring* primitive class type
+     * and false otherwise. Or in other words, is this a class type non-redundantly notated with .val ??
+     */
+    public boolean isValueProjection() {
+        return false;
+    }
+
+    /**
+     * Returns the ClassType representing the primitive value type
+     * of this type, if the class of this type is a primitive class
+     * null otherwise
      */
-    public Type valueProjection() {
+    public ClassType asValueType() {
```

Comment:
These various internal APIs are still crystallizing. I have a follow up ticket JDK-8268734 to evolve them to final form.